yarn-project/prover-client/src/orchestrator/orchestrator.ts

```diff
@@ -276,13 +283,21 @@ export class ProvingOrchestrator {
    * @param provingState - The proving state being worked on
    */
   private async startTransaction(tx: ProcessedTx, provingState: ProvingState) {
-    const txInputs = await this.prepareBaseRollupInputs(provingState, tx);
+    // Pass the private kernel tail vk here as the previous one.
+    // If there are public functions then this key will be overwritten once the public tail has been proven
+    const previousKernelVerificationKey = provingState.privateKernelVerificationKeys.privateKernelCircuit;
+    const txInputs = await this.prepareBaseRollupInputs(provingState, tx, previousKernelVerificationKey);
     if (!txInputs) {
       // This should not be possible
       throw new Error(`Unable to add padding transaction, preparing base inputs failed`);
     }
     const [inputs, treeSnapshots] = txInputs;
-    const txProvingState = new TxProvingState(tx, inputs, treeSnapshots);
+    const txProvingState = new TxProvingState(
+      tx,
+      inputs,
+      treeSnapshots,
+      provingState.privateKernelVerificationKeys.privateKernelToPublicCircuit,
```

Comment:
I guess it doesn't matter based on the comment above, but I'm confused why we use the `ToPublic` keys here, but prepare for the base rollup?

Reply:
So, if there are public functions, then this is the key that will be provided to the first public kernel to verify the previous private kernel. The key that gets given to the base rollup in this case will be the `publicKernelTail` and this is done once that circuit has been proven.

If there no public functions then it uses the `priveKernelCircuit` key provided further up to `prepareBaseRollupInputs`.